### PR TITLE
docs: Add AFL Implementation Document

### PR DESCRIPTION
**Pull request by retrage**
## Type of PR

- [x] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [ ] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [ ] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Related Issue

#19

## Importance of PR

- Importance of the issue
    - [ ] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [x] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [ ] Large (requires several days to several weeks of review)
    - [x] Medium (requires several hours to a day of review)
    - [ ] Small (trivial changes, such as build error)

## PR Overview

#19 

This PR adds afl/implementation_en.md that translated from afl/implementation_ja.md with minor changes. It also deduplicates To-Do notes with TODO.md so that these note is placed under TODO.md.

## Concerns (Optional)

- [ ] Performance
- [ ] Source Code Quality

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [ ] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [x] The PR title is a summary of the changes.
- [x] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [ ] CI is green or confirmed test run results.
- [ ] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [ ] The title of this PR summarizes the changes made by this PR properly.
- [ ] The target branch of this PR is as intended.
- [ ] The reviewer understands the issues in this PR.
- [ ] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

### CODE REVIEW

File: docs/algorithms/afl/implementation_ja.md [hunk 1]
```diff
@@ -45,7 +43,13 @@ fuzzufにおいてAFLは `include/fuzzuf/algorithms/afl` および `algorithms/a
     );
 ```
 
-これらのノードの内、`bit_flip1`, `bit_flip_other`などのノードのルーチンは、`include/fuzzuf/algorithms/afl/afl_mutation_hierarflow_routines.hpp` で宣言されています。`normal_update`, `construct_auto_dict`などのノードのルーチンは、`include/fuzzuf/algorithms/afl/afl_update_hierarflow_routines.hpp` で宣言されています。それらのノード以外のルーチンは、 `include/fuzzuf/algorithms/afl/afl_other_hierarflow_routines.hpp` で宣言されています。
+各ノードのルーチンは次のコードで実装されています:
+- `bit_flip1` and `bit_flip_other`:
```
Comment (potetisensei) [inline docs/algorithms/afl/implementation_ja.md]: I think this wording would bring a confusion (the meaning changes before and after). 
I prefer something like "mutation routines (e.g., `bit_flip1` and `bit_flip_other`)" compared to this.

File: docs/algorithms/afl/implementation_ja.md [hunk 2]
```diff
@@ -9,8 +9,6 @@
 
 fuzzufにおいてAFLは `include/fuzzuf/algorithms/afl` および `algorithms/afl` 以下で実装されています。AFLは派生アルゴリズムを多く持つ性質上、ほとんどの構成クラスがテンプレートによって実装されます。したがって、それらのクラスの宣言・定義は全て `include/fuzzuf/algorithms/afl` 以下のヘッダーファイル内にあります。
 
-特に、それらのテンプレートクラスの宣言は `include/fuzzuf/algorithms/afl` 直下のファイル、定義は `include/fuzzuf/algorithms/afl/templates` 直下のファイルで行われています。これは、明確にfuzzufのコーディング規約で定まっているルールというわけではありません。AFLは、元々通常のクラスとして実装されていたものをテンプレートクラスとして実装し直しており、git上の差分を少なくするためにはそのような構成にせざるをえなかったという歴史的な経緯によります。今後のリファクタリングよって、`templates` サブディレクトリを削除し、ファイルをまとめることを検討しています。
```
Comment (potetisensei) [inline docs/algorithms/afl/implementation_ja.md]: I think this paragraph has some degrees of importance and hence should not be removed.
Actually this is not written in TODO.md, and would be confusing for those who don't know why we split files this way. 

Considering that this `/templates/` sub-directory can be seen in other algorithms than AFL, I agree to remove this paragraph if we move it somewhere else by for example noting about this in TODO.md.

**Review by potetisensei — CHANGES_REQUESTED** [review]
Thank you for letting me know the changes in AFL documents.
I left some comments so could you check them?

**Review by mkYYY — COMMENTED** [review]
#### Mandatory Entries
- [x] The title of this PR summarizes the changes made by this PR properly.
- [x] The target branch of this PR is as intended.
- [x] The reviewer understands the issues in this PR.
- [x] The reviewer plans to review with an appropriate workload based on the importance of this PR.
---

The current translation seems to be accurate. 
I agree @potetisensei 's opinion to keep the paragraph mentioning AFL's `templates` sub-directory somewhere.

File: docs/algorithms/afl/implementation_ja.md (hunk 1 above)
Comment (retrage) [reply docs/algorithms/afl/implementation_ja.md]: Fixed with [fd84959](https://github.com/fuzzuf/fuzzuf/pull/20/commits/fd8495922ac0eb726699c0f7188aab47e7c7b44c).

File: docs/algorithms/afl/implementation_ja.md (hunk 2 above)
Comment (retrage) [reply docs/algorithms/afl/implementation_ja.md:12]: [4edbf1a](https://github.com/fuzzuf/fuzzuf/pull/20/commits/4edbf1a9cf736fb6c34795f3d4975a3f35a17a1f) moves this paragraph to TODO.md.

Comment (retrage) [reply docs/algorithms/afl/implementation_ja.md]: [6377ba0](https://github.com/fuzzuf/fuzzuf/pull/20/commits/6377ba0398b72c412dcfdec395452a9932674a47) reverts this change.

**Review by mkYYY — APPROVED** [review]
#### Mandatory Entries
- [x] The reviewer understands how this PR addresses the issue and the specific changes.
- [x] This PR uses the best possible issue resolution that the reviewer can think of.
- [x] This PR is ready to be merged.

File: docs/algorithms/afl/implementation_en.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+## Reference AFL Implementation
+
+- Version: 2.57b
+- Commit: [fab1ca5ed7e3552833a18fc2116d33a9241699bc](https://github.com/google/AFL/commit/fab1ca5ed7e3552833a18fc2116d33a9241699bc)
+
+## Directory Layout
+
+In fuzzuf, implementations of AFL are located in [include/fuzzuf/algorithms/afl](/include/fuzzuf/algorithms/afl) and [algorithms/afl](/algorithms/afl). Because AFL has many derived algorithms, we implemented most of the classes by templates. Therefore, the declarations and definitions of those classes are in the header files under [include/fuzzuf/algorithms/afl](/include/fuzzuf/algorithms/afl).
+
+These template classes are declared in [include/fuzzuf/algorithms/afl](/include/fuzzuf/algorithms/afl) and defined in [include/fuzzuf/algorithms/afl/templates](/include/fuzzuf/algorithms/afl/templates). This directory layout is due to historical reasons and not due to the coding conventions of fuzzuf; it is designed to reduce the difference of changes when the AFL implementation is reconstructed as template classes to improve reusability. We consider removing the `templates` subdirectory and merging the files in future refactorings.
```
Comment (potetisensei) [inline docs/algorithms/afl/implementation_en.md]: > it is designed to reduce the difference of changes when the AFL implementation is reconstructed as template classes to improve reusability.

it was designed ~ AFL implementation was reconstruted?

**Review by potetisensei — APPROVED** [review]
Great job.

> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [x] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [x] If minor refactorings are not mentioned in the PR, I understand the intent.
- [x] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [x] The reviewer understands how this PR addresses the issue and the specific changes.
- [x] This PR uses the best possible issue resolution that the reviewer can think of.
- [x] This PR is ready to be merged.

**Pull request merged**